internal/action/actions.go

```diff
@@ -160,6 +160,41 @@ func (h *BufPane) Center() bool {
 	return true
 }
 
+// CursorToViewTop moves the cursor to the top of the view, offset by scrollmargin
+func (h *BufPane) CursorToViewTop() bool {
+	v := h.GetView()
+	h.Buf.ClearCursors()
+	scrollmargin := int(h.Buf.Settings["scrollmargin"].(float64))
+	h.Cursor.GotoLoc(h.LocFromVLoc(display.VLoc{
+		SLoc:    h.Scroll(v.StartLine, scrollmargin),
```

Comment:
Overall LGTM, but I still find it rather non-intuitive that we add scrollmargin even if the view is at the beginning of the file. Especially in the case when the cursor itself is at the beginning of the file, i.e. if the cursor already at the very top of the view. In this case `CursorToViewTop` unexpectedly moves the cursor a few lines down.

And similar concerns about `CursorToViewBottom`.

(Also for example in a pathological case when the view height is just 4 lines, `CursorToViewTop` moves the cursor to the bottom, and `CursorToViewBottom` moves the cursor to the top.)

But if you insist, I think we can leave it as is.

Reply:
I can confirm these concerns.

In case we really add it, `CursorToViewTop` & `CursorToViewBottom` should do what they're named for...place the cursor at the top or bottom line. Otherwise there is no big difference to the current behavior expect the reset of the cursor position to the first row.
`CursorToViewCenter` indeed could become handy. For the both others it's currently hard to find the reasons for.
Maybe just one more action bringing the cursor to the start of the line instead of `CursorToViewTop` & `CursorToViewBottom`, which can be chained too?

Reply:
If the actual use case is just to move the cursor one page up or down without changing the cursor's position in the view, then perhaps we should just directly address this use case?

I.e. add new actions similar to `CursorPageUp` and `CursorPageDown` but preserving the cursor position in the view.

Or... perhaps just change the behavior of `CursorPageUp` and `CursorPageDown` themselves, making them preserve the position in the view? By the way, that is nano's behavior of page up/down. And some micro users already expressed a desire to have this behavior in micro: https://github.com/zyedidia/micro/issues/1808.

Something like this:

```go
diff --git a/internal/action/actions.go b/internal/action/actions.go
index cf6d954f..05b7a15e 100644
--- a/internal/action/actions.go
+++ b/internal/action/actions.go
@@ -1657,6 +1657,9 @@ func (h *BufPane) CursorPageUp() bool {
 		h.Cursor.StoreVisualX()
 	}
 	h.MoveCursorUp(h.BufView().Height)
+
+	h.ScrollUp(h.BufView().Height)
+	h.ScrollAdjust()
 	h.Relocate()
 	return true
 }
@@ -1671,6 +1674,9 @@ func (h *BufPane) CursorPageDown() bool {
 		h.Cursor.StoreVisualX()
 	}
 	h.MoveCursorDown(h.BufView().Height)
+
+	h.ScrollDown(h.BufView().Height)
+	h.ScrollAdjust()
 	h.Relocate()
 	return true
 }
```

(and same for `SelectPageUp` and `SelectPageDown`)

I don't think I ever particularly liked micro's current behavior, althought I haven't really thought about it, I think I've always been subconsciously irritated by it, and would enjoy nano-style behavior much more. (And I suppose micro's current behavior is what it is not because it was actually preferred by @zyedidia or anyone, but just because it was the quickest to implement.)

Reply:
> Or... perhaps just change the behavior of `CursorPageUp` and `CursorPageDown` themselves, making them preserve the position in the view? By the way, that is nano's behavior of page up/down.

Also `gnome-text-editor` does the same (and both preserve the cursor position within the line too). And by no I realized, that `micro` doesn't...
So good point, but in the moment we change the default behavior `CursorPageUp` & `CursorPageDown` we break with the behavior someone might become used to it.
So have we any other chance than implementing new actions?

Reply:
I'd rather change the behavior of `CursorPageUp` & `CursorPageDown` and see if we get any complaints. If we get none, no problem. If we get some, we can revert it and add new actions, at least knowing that it makes sense to do so.

Although there is a high chance that if we get complaints, we get them only after the next release, but, well, what can we do.

Reply:
> and both preserve the cursor position within the line too

Micro preserves it too, actually.

Reply:
> Overall LGTM, but I still find it rather non-intuitive that we add scrollmargin even if the view is at the beginning of the file. Especially in the case when the cursor itself is at the beginning of the file, i.e. if the cursor already at the very top of the view. In this case `CursorToViewTop` unexpectedly moves the cursor a few lines down.

Yes, I don't like that behavior either, but as I mentioned earlier, it's the only way to avoid an unexpected jump if the user has `scrollmargin` set. To recap: if `scrollmargin` is set to 5, and the cursor is moved to the top, when the user hits an arrow key, the view jumps by 5 lines so that the cursor is 5 lines from the top of the view.

Reply:
> To recap: if `scrollmargin` is set to 5, and the cursor is moved to the top, when the user hits an arrow key, the view jumps by 5 lines so that the cursor is 5 lines from the top of the view.

Ah, that's what you meant. I thought you were talking about a different thing.

So, what I was talking about: if the *view* itself is already at the top of the file, then if the cursor is moved to the top and the user hits an arrow key, the view does *not* jump by 5 lines. Because it is at the top of the file, not just at the top of the view.

So your implementation is inconsistent with that: it offsets by scrollmargin even when there is no need for that, since the view is at the top of the file.

...Anyway, what do you think about simply changing the behavior of `CursorPageUp` and `CursorPageDown` per #1808, instead of adding all these actions?

Reply:
I'm very much in favor of changing the `CursorPageUp` and `CursorPageDown` actions to be consistent with other editors (as described, but there would still be a need for these new actions. The use cases are different:

- The new `CursorPage[Up/Down]` actions would always move the view up or down by a page, leaving the cursor where it was relative to the view (except when the cursor would be placed beyond the end of the file, in which case it would be moved to the last line of the file). We need to define exactly how `scrollmargin` would be handled here, because nano does not implement a scrollmargin, AFAIK.
- The `CursorToView[Top/Center/Bottom]` actions always only move the cursor, without changing the view by even one line. Other editors provide these actions (for instance, helix) and I'm sure people find them useful.

Reply:
> The use cases are different: [...]

You've described the behaviors, not the use cases for them. :)

But ok, I trust you that you have some uses for `CursorToView[Top/Center/Bottom]`.

But still, as I said, I believe the `CursorToView[Top/Bottom]` implementation should be corrected to take into account whether the view itself isn't at the very top/bottom, i.e. whether it makes sense to offset by scrollmargin.

Or perhaps, like @JoeKar suggested, just add `CursorToViewCenter` only, leaving out `CursorToView[Top/Bottom]` (thus keep it simple and avoid users confusion by this non-intuitive scrollmargin-induced behavior). Unless you find `CursorToView[Top/Bottom]` particularly useful too?

Reply:
I think that if we implement a `CursorToViewCenter` action then people would probably expect to have `CursorToViewTop` and `CursorToViewBottom` as well? They're useful, anyway, especially if you have a large monitor.

I've added logic to not add scrollmargin when the view is at the beginning or end of the file.